pkg/client/fake/client.go

```diff
@@ -290,9 +293,22 @@ func (t versionedTracker) Add(obj runtime.Object) error {
 		if err != nil {
 			return fmt.Errorf("failed to get accessor for object: %w", err)
 		}
-		if accessor.GetDeletionTimestamp() != nil && len(accessor.GetFinalizers()) == 0 {
-			return fmt.Errorf("refusing to create obj %s with metadata.deletionTimestamp but no finalizers", accessor.GetName())
+
+		gvk, err := apiutil.GVKForObject(obj, t.scheme)
+		if err != nil {
+			panic(fmt.Errorf("failed to get gvk for object %T: %w", obj, err))
+		}
+
+		if !allowsGracefulDelete(gvk, obj) {
+			if accessor.GetDeletionTimestamp() != nil && len(accessor.GetFinalizers()) == 0 {
+				return fmt.Errorf("refusing to create obj %s with metadata.deletionTimestamp but no finalizers", accessor.GetName())
+			}
+		} else {
+			if accessor.GetDeletionTimestamp() != nil && pointer.Int64Deref(accessor.GetDeletionGracePeriodSeconds(), 0) <= 0 {
```

Comment:
When `DeletionGracePeriodSeconds` is non-zero it means the Pod can be observed because it will be deleted after that grace period. That means one can see the deleting Pod in the time range `[deleted timestamp, deleted timestamp + grace period seconds)`. 

> From what I can tell it allows the value to be negative and will then set it to 0 when executing the deletion

I think it will be set to 1.

```go
	if period < 0 {
		period = 1
	}
```